x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -387,4 +420,219 @@ public void usedDeprecatedField(String usedName, String replacedWith) {
                 usedName, apiKeyId, replacedWith);
         }
     }
+
+    /**
+     * Invalidate API keys for given realm and user name.
+     * @param realmName realm name
+     * @param userName user name
+     * @param invalidateListener listener for {@link InvalidateApiKeyResponse}
+     */
+    public void invalidateApiKeysForRealmAndUser(String realmName, String userName,
+            ActionListener<InvalidateApiKeyResponse> invalidateListener) {
+        ensureEnabled();
+        if (Strings.hasText(realmName) == false && Strings.hasText(userName) == false) {
+            logger.trace("No realm name or username provided");
+            invalidateListener.onFailure(new IllegalArgumentException("realm name or username must be provided"));
+        } else {
+            findActiveApiKeysForUserAndRealm(userName, realmName, ActionListener.wrap(apiKeyIds -> {
+                    if (apiKeyIds.isEmpty()) {
+                        logger.warn("No api keys to invalidate for realm [{}] and username [{}]", realmName, userName);
+                        invalidateListener.onResponse(InvalidateApiKeyResponse.emptyResponse());
+                    } else {
+                        invalidateAllApiKeys(apiKeyIds, invalidateListener);
+                    }
+                }, invalidateListener::onFailure));
+        }
+    }
+
+    private void invalidateAllApiKeys(Collection<String> apiKeyIds, ActionListener<InvalidateApiKeyResponse> invalidateListener) {
+        indexInvalidation(apiKeyIds, invalidateListener, null);
+    }
+
+    /**
+     * Invalidate API keys for given API key id
+     * @param apiKeyId API key id
+     * @param invalidateListener listener for {@link InvalidateApiKeyResponse}
+     */
+    public void invalidateApiKeysForApiKeyId(String apiKeyId, ActionListener<InvalidateApiKeyResponse> invalidateListener) {
+        ensureEnabled();
+        invalidateAllApiKeys(Collections.singleton(apiKeyId), invalidateListener);
+    }
+
+    /**
+     * Invalidate API keys for given API key name
+     * @param apiKeyName API key name
+     * @param invalidateListener listener for {@link InvalidateApiKeyResponse}
+     */
+    public void invalidateApiKeysForApiKeyName(String apiKeyName, ActionListener<InvalidateApiKeyResponse> invalidateListener) {
```

Comment:
This comment was a better fit for https://github.com/elastic/elasticsearch/pull/36029 but I missed it :) I think we should not allow multiple API keys with the same name. The semantics are for us to define but the fact that we require the `name` to be set, and that we would allow to invalidate by `name` makes it look like an id. As a user, I think I would expect to overwrite my api key when I create a new one with the same name, or get a warning that one exists already.

Reply:
> WDYT of this behavior?

Agreed :)

>. If the key by the same name has already been invalidated, then the creation of new API key with the same name would be allowed

 I was wondering if it's worth adding a message to that effect in the error we return, but while typing this out. it makes more sense to add these details in the documentation instead. I'll just leave this comment as a reminder for that.